### ISSUE TITLE
could you please share your test data

### ISSUE THREAD

**Issue by degilala** [Algolzw/SPAC-Deformable-Registration#8]
could you please share your test data lspig_affine_test_3d.h5 and sliver_affine_test_3d.h5. Thank you so much

**Comment by degilala** [comment]
I've already seen your response in other posts，Thank you very much

**Comment by Algolzw** [comment]
No worries,

Best,
Ziwei

**Comment by degilala** [comment]
There is another question。When I reproduced your code，I used Lits as the training data。 But the reward seems to have some problem：
------------------------Saving model------------------
------------------------Model saved!------------------
ep-4900- 7:  -- critic: 94.0003, alpha: 0.2908, reg: 97.2448, reward: 0.408, score: 0.891, init_score: 0.794
ep-4900-15:  -- critic: 91.8634, alpha: 0.2908, reg: 102.3696, reward: 0.096, score: 0.906, init_score: 0.794
ep-4910- 7:  -- critic: 75.0961, alpha: 0.2907, reg: 116.8783, reward: -0.043, score: 0.768, init_score: 0.714
ep-4910-15:  -- critic: 51.6687, alpha: 0.2907, reg: 114.6242, reward: -0.333, score: 0.751, init_score: 0.714
ep-4920- 7:  -- critic: 139.8706, alpha: 0.2906, reg: 92.1535, reward: -0.050, score: 0.914, init_score: 0.843
ep-4920-15:  -- critic: 465.4319, alpha: 0.2905, reg: 114.6815, reward: -0.192, score: 0.900, init_score: 0.843
ep-4930- 7:  -- critic: 83.6345, alpha: 0.2904, reg: 109.2419, reward: 0.382, score: 0.898, init_score: 0.805
ep-4930-15:  -- critic: 63.8499, alpha: 0.2904, reg: 127.2785, reward: -0.100, score: 0.902, init_score: 0.805
ep-4940- 7:  -- critic: 50.2921, alpha: 0.2903, reg: 114.1414, reward: -0.121, score: 0.887, init_score: 0.810
ep-4940-15:  -- critic: 292.7428, alpha: 0.2902, reg: 112.0293, reward: -0.223, score: 0.874, init_score: 0.810
ep-4950- 7:  -- critic: 203.1610, alpha: 0.2901, reg: 101.3863, reward: 0.027, score: 0.904, init_score: 0.840
ep-4950-15:  -- critic: 107.9155, alpha: 0.2901, reg: 154.2123, reward: -0.102, score: 0.900, init_score: 0.840
ep-4960- 7:  -- critic: 128.4205, alpha: 0.2900, reg: 112.3422, reward: 0.020, score: 0.852, init_score: 0.801
ep-4960-15:  -- critic: 77.1781, alpha: 0.2900, reg: 123.1816, reward: -0.060, score: 0.853, init_score: 0.801
ep-4970- 7:  -- critic: 78.9349, alpha: 0.2898, reg: 113.9834, reward: 0.139, score: 0.839, init_score: 0.763
ep-4970-15:  -- critic: 87.4864, alpha: 0.2898, reg: 102.6824, reward: -0.078, score: 0.837, init_score: 0.763
ep-4980- 7:  -- critic: 149.0953, alpha: 0.2897, reg: 94.4010, reward: 0.038, score: 0.769, init_score: 0.667
ep-4980-15:  -- critic: 432.5048, alpha: 0.2897, reg: 124.2985, reward: -0.351, score: 0.751, init_score: 0.667
ep-4990- 7:  -- critic: 60.5094, alpha: 0.2895, reg: 110.8110, reward: 0.298, score: 0.813, init_score: 0.731
ep-4990-15:  -- critic: 504.1006, alpha: 0.2895, reg: 109.8992, reward: -0.086, score: 0.813, init_score: 0.731
The reward is always negative and the moved in process seems very strange. I didn't change any parameters.Could you please help me ? Thank you so much!By the way Does the parameters in training 3d data are seem as training 2d data?

**Comment by Algolzw** [comment]
The shown reward is produced in the last step of an episode, it is indeed possible to be negative. But from init_score to the current score it seems images are registered well. You can train the agent for more episodes until the learning is stable. 

The parameter settings of 3D data are similar to 2D, but you may want to tune the "Done threshold score" to fit 2D image registration.

**Comment by degilala** [comment]
非常感谢您能回复我的消息，我在训练过程中的process文件夹下的moved18.bmp看起来有点奇怪，不知道是不是因为我训练的episcode不够，另外我还想请问一下您训练时需要的内存是多大，我是使用云服务器进行训练的，内存60g，在训练到5500的时候内存就被占满了。
fixed：![fixed](https://github.com/degilala/registration-/blob/main/fixed.PNG)
moved18:![moved18](https://github.com/degilala/registration-/blob/main/moved.PNG)
log曲线:![log_reward](https://github.com/degilala/registration-/blob/main/reward.PNG)
(这里的moved18大概是3000epsiode左右的截图)
再次感谢您的回复，谢谢谢谢

**Comment by degilala** [comment]
另外我还有一个疑惑，在test.py文件里的 
    brain.load_model('actor', cfg.ACTOR_MODEL)
    brain.load_model('decoder', cfg.DECODER_MODEL_RL)
    brain.load_model('critic1', cfg.CRITIC1_MODEL)
    brain.load_model('critic2', cfg.CRITIC2_MODEL)
我在config.py文件中没找到DECODER_MODEL_RL，只有ENCODER_MODEL。不好意思打扰了

**Comment by Algolzw** [comment]
你可以设置小一点的memory replay，32G内存应该就可以。除此之外也可以把它保存到硬盘里(比如append到txt文件里)，这样train会慢一点，但不会有内存限制。

ENCODER_MODEL可能是名字写错了，应该改一改就行。

**Comment by degilala** [comment]
好好，谢谢您，非常感谢

**Comment by degilala** [comment]
您好，不好意思又来打扰您。
首先，关于2D脑图的训练，我看您在其他帖子中的回复，可以下载RCN所使用的数据集，但是RCN使用的脑数据集是3d的，在brain.json中的image_size也是（128，128，128），请问您能分享2d数据集嘛，您2d脑数据集训练使用了三种mask计算dice嘛？
另外，我还想请问一下在config.py中的TRAIN_DATA和TRAIN_SAG_DATA的含义，TRAIN_SAG_DATA是用来训练输入的mask嘛？main.py中的datasets = BrainData(cfg.TRAIN_DATA, cfg.TRAIN_SEG_DATA, cfg.ATLAS, use_seg=cfg.USE_SEG, mode='train', aug=cfg.BSPLINE_AUG)是用来加载2d脑配准的嘛？
还有，我在加载liits.h5的肝脏数据集时，fix_seg和mov__seg是只有一种分割mask嘛，我保存了一下读进来的mask，它看起来是下面这样，我不是非常理解dice计算奖励这里。
![mask](https://github.com/degilala/registration-/blob/main/mask'.PNG)
最后，我最终是想用该代码训练我自己的3dADNI数据集，关于计算reward，我是否可以使用分割软件分割WM，CSF，GM三类mask作为输入，如果使用多种mask计算reward，那我该怎么处理dice的值呢，因为每种mask都会计算出来一个dice。
真的不好意思一再打扰您

**Comment by Algolzw** [comment]
Hello，2D数据集我这里是没有了，不过一般处理流程是先对Brain3D数据进行affine对齐然后对大脑进行切片也就是slices。这里的mask都是用Kmeans来生成的，我设置的是3类，不过也可以按需求变化。

至于肝脏类的3D数据，他们的mask都只有一类，一般就是中心处整个肝脏部分。条件允许的话当然是可以使用其他软件分割计算mask的。多label的dice score用utils的[dice function](https://github.com/Algolzw/SPAC-Deformable-Registration/blob/main/code/utils.py#L67)应该是可以算的吧。

**Comment by degilala** [comment]
好的好的，谢谢您

**Comment by degilala** [comment]
您好，我在训练过程中得到的warp是正常的，但是调用模型进行测试时，得到的warp图是错误的色块，测试代码如下：
def get_register_dataloder(test_dir):
    train_transform = transforms.Compose([transforms.ToTensor()])
    reg_data = myDataset.pair_Dataset(data_dir=test_dir, transform=train_transform)
    #reg_data = lung_iowa_list.LungDataset._get_file_names(self)
    reg_loader = DataLoader(dataset=reg_data, batch_size=1)
    assert len(reg_loader) > 0, 'Could not find any training data'
    return reg_data, reg_loader
if __name__ == "__main__":
    idx = 100
    setup_seed(cfg.SEED)
    utils.remkdir(cfg.TEST_PATH)

    #######################################
    # stn = SpatialTransformer(device).to(device)
    stn = SpatialTransformer(cfg.HEIGHT).to(device)
    seg_stn = SpatialTransformer(cfg.HEIGHT, mode='nearest').to(device)
    reg_data, reg_loader = get_register_dataloder((r'F:\peizhun\SPAC-Deformable-Registration-main\mydata\128', r'F:\peizhun\SPAC-Deformable-Registration-main\mydata\mov_128'))
    brain = SPAC(stn, device)

    brain.load_model('planner', cfg.ACTOR_MODEL)
    brain.load_model('actor', cfg.ACTOR_MODEL_RL)
    brain.load_model('critic1', cfg.CRITIC1_MODEL)
    brain.load_model('critic2', cfg.CRITIC2_MODEL)

    brain.eval(brain.actor)
    brain.eval(brain.critic1)
    brain.eval(brain.critic2)
    brain.eval(brain.planner)
    dices = []
    times = []
    # grid = utils.virtual_grid(cfg.HEIGHT, tensor=True).unsqueeze(0).to(device)
    # print(grid.shape)
    print(len(reg_loader))
    latents = []
    labs = []
    jacobian_dets = []
    for i, item in enumerate(reg_loader):
        fixed = item[0]
        moving = item[1]
        fixed = fixed.to(device)
        moving = moving.to(device)
        moved = copy.deepcopy(moving)

        pred = None
        best_pred = None
        best_value = 0
        best_step = 0
        step = 0
        tic = time.time()
        while step < 20:
            state = torch.cat([fixed, moved], dim=1)
            latent, flow = brain.choose_action(state, test=True)
            latents.append(latent.cpu().numpy())
            labs.append(i)
            pred = flow if pred is None else stn(pred, flow) + flow
            moved = stn(moving, pred)
            step += 1
        toc = time.time()
        wrap_save_path=r"F:\peizhun\SPAC-Deformable-Registration-main\mydata\moved"+"/"+"moved"+str(i)+".nii"
        warped_im = stn(moving, pred).cpu().detach().numpy()
        fixed_path = r'F:\peizhun\SPAC-Deformable-Registration-main\mydata\128\mni_icbm152_t1_tal_nlin_sym_09a_brain.nii'
        fixed_sitk = itk.ReadImage(fixed_path)
        warp_ori = itk.GetImageFromArray(warped_im[0,0,:,:,:])
        warp_ori.SetDirection(fixed_sitk.GetDirection())
        warp_ori.SetOrigin(fixed_sitk.GetOrigin())
        warp_ori.SetSpacing(fixed_sitk.GetSpacing())
        itk.WriteImage(warp_ori, wrap_save_path)
我是使用自己的方式调用.nii数据，调用的.ckpt文件确认没问题，测试数据集和训练数据集是一样是，但我不知道问题出在哪里了，请问您能帮帮我嘛，谢谢谢谢

**Comment by degilala** [comment]
我尝试使用1000_ckpt的模型测试，结果比13000_.ckpt的结果好一点，但仍然是错误的。而训练过程中训练到13000时候保存下来的结果图是正确的，我也尝试取消step叠加形变场进行测试，结果比叠加20次的结果正常一点。我不明白问题出现在哪里了，您有时间的话能帮帮我嘛，不好意思再次打扰您。
这是测试代码[test](https://github.com/degilala/test/upload)

**Comment by Algolzw** [comment]
你好，结果图是什么样的呢？你能发一下正常图和结果图的对比吗

**Comment by degilala** [comment]
参考图：[fix](https://github.com/degilala/test/blob/main/fix.PNG) 浮动图：[mov](https://github.com/degilala/test/blob/main/mov.PNG)
测试过程中只进行一次形变[moved_withoutstep](https://github.com/degilala/test/blob/main/mov_withoutstep.PNG)
测试过程中使用step逐步形变（我展示的test代码的结果）：[mov_withstep](https://github.com/degilala/test/blob/main/moved_withstep.PNG)
训练过程保存的结果（左上角是形变图，右边是浮动图,左下角是参考图）[train](https://github.com/degilala/test/blob/main/train.PNG)

**Comment by degilala** [comment]
我感觉测试时进行一次形变配准效果不如训练时的结果，但进行逐步形变结果会开始过分扭曲
这是进行step等于2时的结果[step2](https://github.com/degilala/test/blob/main/step2.PNG)
这是进行step等于2时的.nii文件[step2nii](https://github.com/degilala/test/blob/main/step2ADNI_002_S_0295_MR_MPRAGE_br_raw_20110602125255238_39_S110476_I238627_brain.nii236.nii)
这是进行一次形变的moved.nii文件[step1](https://github.com/degilala/test/blob/main/ADNI_002_S_0295_MR_MPRAGE_br_raw_20110602125255238_39_S110476_I238627_brain.nii236.nii)
这是fix和mov的.nii文件[fix](https://github.com/degilala/test/blob/main/mni_icbm152_t1_tal_nlin_sym_09a_brain.nii) [mov](https://github.com/degilala/test/blob/main/ADNI_002_S_0295_MR_MPRAGE_br_raw_20110602125255238_39_S110476_I238627_brain.nii)
这是训练时保存结果的moved18和moving的.nii文件[moved18](https://github.com/degilala/test/blob/main/18moved.nii) [moving](https://github.com/degilala/test/blob/main/18moving.nii)

**Comment by Algolzw** [comment]
看了一下图，感觉是一开始flow预测就出问题了，之后多步预测加重了flow的问题。你的训练集和测试集差别大吗？比如分辨率什么的？或者你可以试一试使用训练集的图像进行测试能对齐吗？

你的测试代码没看出来什么问题。你也可以试一试把 ` latent, flow = brain.choose_action(state, test=True)`中改成`test=False`.

**Comment by degilala** [comment]
我现在的测试集和训练集就是同一个数据集，test改成等于false结果也不好，这是我全部的代码[spac](https://github.com/degilala/test/tree/main/spac)
另外，我在训练的时候score一直没有明显的涨幅，不过reward曲线看起来还是正常的
log图：[log](https://github.com/degilala/test/blob/main/log.PNG)

**Comment by degilala** [comment]
您好，我找到问题在哪了，我想问您一下您在训练数据的时候灰度归一化范围是多少啊？谢谢谢谢

**Comment by Algolzw** [comment]
> 您好，我找到问题在哪了，我想问您一下您在训练数据的时候灰度归一化范围是多少啊？谢谢谢谢

Nice! 训练测试应该都是0-1之间吧。
